### DISCUSSION TITLE
ThreadX Examp[le

### DISCUSSION THREAD

**Discussion by hartgordon** [MaJerle/lwesp#117] [Q&A] [answered]
Is there an STM32 / ThreadX example available?

Best Regards
Gordon

**Comment by MaJerle** [comment, marked as answer]
Not public for the moment - I have it on my private project hence you see some drivers here.
Will plan to make it. Could you please open an issue?

**Reply by hartgordon** [reply]
I have done so, thanks.  In the meantime I have managed to get a solution together with one exception.  Could you share assist on how to configure the memory management under ThreadX:

void*   lwesp_mem_malloc(size_t size);
void*   lwesp_mem_realloc(void* ptr, size_t size);
void*   lwesp_mem_calloc(size_t num, size_t size);
void    lwesp_mem_free(void* ptr);
uint8_t lwesp_mem_free_s(void** ptr);

Thanks!

**Comment by MaJerle** [comment]
LwESP lib has some sort of ThreadX port. It has system functions (OS) and memory functions. See here for instance: https://github.com/MaJerle/lwesp/blob/develop/lwesp/src/system/lwesp_mem_threadx.c

You will find "extern" variable for memory pool. Idea is that LwESP port part for THreadX uses memory pool defined by user from the application level.

Init sequence would be then:

```c
/* Init your TX memory pool */

    /* Set byte pool for memory operations */
    lwesp_sys_preinit_threadx_set_bytepool_handle(&tx_app_byte_pool);

    /* Init stack, and send reset/restore commands in non-blocking mode */
    if (lwesp_init(lwesp_main_evt_fn, 0) != lwespOK) {
        debug_printf("[LwESP] Could not initialize stack\r\n");
        return 0;
    }
    debug_printf("[LwESP] Stack initialized\r\n");

```
`lwesp_sys_preinit_threadx_set_bytepool_handle` exists only for ThreadX port. In my case, I init TX byte pool in `tx_application_define` callback function

Second problem of ThreadX is that you cannot terminate yourself (cleanup self thread, like in FreeRTOS). For that sense you need to have another thread that does it for you - usually called Idle Thread. Not having idle thread is OK for core LwESP threads to run, but you cannot have multiple application threads that are started and terminated on the fly, several times. That's the only drawback.
